spoon-smpl/src/test/java/spoon/smpl/MetavarsTest.java

```diff
@@ -0,0 +1,145 @@
+package spoon.smpl;
+
+import fr.inria.controlflow.ControlFlowNode;
+import org.junit.Before;
+import org.junit.Test;
+import spoon.smpl.formula.*;
+import spoon.smpl.metavars.ConstantConstraint;
+import spoon.smpl.metavars.ExpressionConstraint;
+import spoon.smpl.metavars.IdentifierConstraint;
+import spoon.smpl.metavars.TypeConstraint;
+
+import java.lang.reflect.Field;
+import java.util.*;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+import static spoon.smpl.TestUtils.*;
+import static spoon.smpl.TestUtils.parseReturnStatement;
+
+/**
+ * This is essentially an integration test of most of the SmPL stack, leaving out the parser.
+ */
+public class MetavarsTest {
+    @Before
+    public void resetControlFlowNodeCounter() {
+        // needed for consistent IDs in CFGModels
+        try {
+            Field field = ControlFlowNode.class.getDeclaredField("count");
+            field.setAccessible(true);
+            ControlFlowNode.count = 0;
+        } catch (Exception e) {
+            fail("Unable to reset ControlFlowNode id counter");
+        }
+    }
+
+    private static Formula stmt(String code, Map<String, ParameterPostProcessStrategy> metavars) {
+        return new StatementPattern(makePattern(parseStatement(code), new ArrayList<>(metavars.keySet())), metavars);
+    }
+
+    private static Formula retstmt(String code, Map<String, ParameterPostProcessStrategy> metavars) {
+        return new StatementPattern(makePattern(parseReturnStatement(code), new ArrayList<>(metavars.keySet())), metavars);
+    }
+
+    @Test
+    public void testCompatibleBindings() {
+
+        // contract: metavariables bound to "same thing" in different nodes are joined under AND
+        // TODO: add tests for other formula connectors such as OR
+
+        Model modelA = new CFGModel(methodCfg(parseMethod("int m() { int x = 1; return x; }")));
+        ModelChecker checkerA = new ModelChecker(modelA);
+        //System.out.println(((CFGModel) modelA).getCfg().toGraphVisText());
+
+        Map<String, ParameterPostProcessStrategy> meta = metavars("z", new IdentifierConstraint());
+
+        stmt("int z = 1;", meta).accept(checkerA);
+        assertEquals("[(4, {z=x})]", checkerA.getResult().toString());
+
+        retstmt("return z;", meta).accept(checkerA);
+        assertEquals("[(5, {z=x})]", checkerA.getResult().toString());
+
+        new And(stmt("int z = 1;", meta),
```

Comment:
it seems that we have two contracts here. what about splitting this test in two: the first two assertions, and then the final one (contract on And)?

Reply:
I like seeing the first two assertions there since the `And` essentially joins those two results, but I agree it distracts from the contract and the repetition is bad. I'll try to restructure.

Reply:
perfect